### PR TITLE
Arrival Page Fixes

### LINKED ISSUE #1412: [BUG] Missing a single STAR in the A320 MCDU but not from FS2020 or other Aircraft
**Mod Version**

built: 2020-10-16T00:15:31+00:00
ref: refs/heads/master
sha: e08f84b3c7db7767aab2c34c8a90ea059bd1f9fe
actor: lukecologne
event_name: push

**Describe the bug**

Some airports are missing a single STAR. 

For example KCMH has 3 STARS that appear in the FS2020 Menu and on other aircraft. The MCDU on the A320 only displays two of the three of them. 

**To Reproduce**

1. Load KCMH as the destination airport in the MCDU
2. Load the "Arrival to KCMH" page under F-PLN
3. Press the right arrow

**Expected behavior**

BREMN5, GAILL3, and GUNNE4 should appear at selectable STARS. 

Pressing the up arrow should reveal the third STAR. 

**Actual behavior**

BREMN5 and GAILL3 appear. GUNNE4 does not appear. Up arrow does nothing. 

**References**

From Game Menu:
image

From A320 (up and down arrow does nothing)
image

From TBM
image

**Additional context**

Possibly related to #1020 which was partially resolved by #1027 

Prior to #1027 no STARs would show up at KCMH, now 2 of the 3 show up. 

I worked with @NathanInnes to determine this appears to be a mod issue. He was able to recreate this issue at another airport. 

Was this working before/when did the issue start occurring?
Has not worked before, as mentioned above prior to #1027 no STARs would show up at KCMH, now 2 of the 3 show up. 

Is this a problem in the vanilla unmodded game?
No STARs appear at all in the vanilla game in the A320. However the STARS do appear on other aircraft and in the menu. 

Discord username (if different from GitHub):
COVID_19_FOX3

### LINKED ISSUE #1436: [BUG] MCDU Arrivals page - Page scroll constrained by length of left list (arrival/approach selection), ignoring right list (transition selection)
**Mod Version**

Dev version:
built: 2020-10-16T00:15:31+00:00
ref: refs/heads/master
sha: e08f84b3c7db7767aab2c34c8a90ea059bd1f9fe
actor: lukecologne
event_name: push

**Describe the bug**

The Arrival Page scroll up/down is limited by the left list (which lists available arrivals and approach procedures), not by the max of that and the right list (available transitions for selected procedure). As a result, **if the number of transitions exceeds the number of available procedures, transitions further down the list can't be selected**

**To Reproduce**

1. Program a flight plan in the MCDU from KSLC to KSNA
2. Program in the DSNEE4 arrival
3. Try to select the MARUE transition for the arrival by scrolling down the list

**Expected behavior**

Ability to scroll down the list of transitions continues. List on left scrolls up and off the page.

**Actual behavior**

Can't continue scrolling because the left list is at its end.

**References**

Can't scroll past TANDY5 arrival:
image

**Additional context**

Was this working before/when did the issue start occurring?
I think this has always been an issue?

Is this a problem in the vanilla unmodded game?
Probably. It may explain the numerous reports of missing transitions to certain approaches.

Discord username (if different from GitHub):
Eefoe#0570

### LINKED ISSUE #2222: [BUG] Last STAR not shown/selectable on MCDU F-PLN ARRIVAL page
**Mod Version**
Master, version #3330179

**Describe the bug**
When selecting a STAR on the arrival page in F-PLN, the last STAR (as ordered alphabetically) isn't shown and cannot be selected.

**To Reproduce**
1. Using the in-game flight planner, note whatever arrival is at the bottom of the list of the destination airport

last_star_ig

_LZARD6 in the case of Tampa, KTPA_

2. Load in at origin airport (C&D in my case), enter origin/dest, go to F-PLN, setup arrival

3. Attempt to scroll beyond the second-to-last STAR on the list in order to select the arrival from step 1

last_star_mcdu
_FOOXX5 is the last option available while LZARD6 should be beneath it_

**Expected behavior**
All available STARs should be displayed and selectable.

**Actual behavior**
The lowest alphabetically-sorted STAR cannot be used

**Additional context**
I'm guessing a for-loop isn't accounting for the added "NO STAR" option in the list.

Approaches for me have also become very glitchy in the past day or so. MCDU not automatically setting the ILS localizer freq, AP aligning with the localizer itself and not the runway (yet again), and the AP ignoring STARs altogether by just going direct to the runway.

I don't think these are actually related issues, but it may or may not come up in testing as well.

### PR DESCRIPTION
**Pull request by saintwolf**
Fixes #1412
Fixes #1436
Fixes #2222

## Summary of Changes
* Fixed duplicate STARs showing for some Airports. In some cases (Such as EHAM), selecting ILS 08 would cause duplicate STARS on the arrivals list. This would also happen when selecting ILS 18. This fix checks to ensure the runway is exact (and also adds a check for Center runways if they exist at the airport).

* Fixed #1412. The last STAR on the list will now be scrollable. This also fixes #1436

* Approach Via page will now stop scrolling at the correct place. Previously the list would keep scrolling down, ad infinitum. 

* Arrival and departure pages will now display the selected Arrivals, Departures, Transitions, etc in yellow if there is a temporary flight plan, otherwise they will display in green.

* Approaches are now ordered as they should be on a Honeywell FMS.

* Arrival and departure pages both show runway course and length.

* Departure page runway list will stop scrolling in the correct place.

### **Before:**
image

image
### **After:**
image

image

### **Runway selection list changes:**
image

image

## References

image

## Additional context

Discord username (if different from GitHub): Saintwolf#5557

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Review by Benjozork — APPROVED** [review]
LGTM

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AvailableArrivalsPage.js [hunk 1]
```diff
@@ -54,7 +54,8 @@ class CDUAvailableArrivalsPage {
                             for (let j = 0; j < arrival.runwayTransitions.length; j++) {
                                 const runwayTransition = arrival.runwayTransitions[j];
                                 if (runwayTransition) {
-                                    if (runwayTransition.name.indexOf(selectedRunway) != -1) {
+                                    // Check if selectedRunway matches a transition on the approach (and also checks for Center runways)
+                                    if (runwayTransition.name.match("^RW" + selectedRunway + "(C)*$")) {
```
Comment (beheh) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AvailableArrivalsPage.js:58]:
```suggestion
                                    if (runwayTransition.name.match("^RW" + selectedRunway + "C?$")) {
```

Can be simplified to this (unless you want to match `RW09CCCCC`).

**Comment by Highflyer2010** [comment]
Category: QA Tester
Name: Discord User Highflyer2010 #8932
Date of testing: Current date 07/12/2020
Version of the sim: Version the tester uses 1.11.6.0
PR Tested: #2263 
Object of test: Checking Douplicate Stars showing at some airports (Without Navigraph)

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:

1. spawned at EGNT 
2. setting up a flight plan to EHAM
3. checking for duplicate entries ILS 08
4. No entry for ILS 08
5. Selecting ILS 18 - no duplicate entries

Flight notes (If required):

Issues (If Any): 
Overall Rating: Good
Conclusions of the testing: Make a small descriptive to areas where it needs to be improved and recommendations.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AvailableArrivalsPage.js (hunk 1 above)
Comment (saintwolf) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AvailableArrivalsPage.js:58]: Yeah this is much better. Tested it through a regex tester and looks like it works just as well :)

**Comment by saintwolf** [comment]
@wpine215 Have added a Changelog

**Comment by beheh** [comment]
Not required, but if you're touching the page anyway and still want to do this, one small change you could do is to update the button labels from the Thales style `TMPY ERASE` / `TMPY INSERT` to the Honeywell-style `{INSERT` / `ERASE*`. Here's a reference:

image

**Comment by saintwolf** [comment]
@beheh No problem! It looks like the Thales style is used in a few files: 
- A320_Neo_CDU_AvailableArrivalsPage.js
- A320_Neo_CDU_AvailableDeparturesPage.js
- A320_Neo_CDU_DirectToPage.js
- A320_Neo_CDU_FlightPlanPage.js

Should probably update them all for consistency. Should I do them all here or create a new PR?

**Comment by beheh** [comment]
I think some of it is just still from Asobo, and we weren't Honeywell from day 1 either. The Flight Plan page is already being worked on in #2245, but the others would be great, thanks!

**Comment by saintwolf** [comment]
@beheh All done!

image
image
image

**Comment by saintwolf** [comment]
Made some more changes to align the departures page with the Honeywell design: 
image

Reference: https://www.theairlinepilots.com/forumarchive/a320/a320-fms-pages.pdf

**Comment by beheh** [comment]
If you want to go the extra mile for accuracy, add a few non-breaking spaces before some of the labels:

image

They're a bit hard to spot at first, but you can't un-see them once you realize they're there :-)
You can use `{sp}` [like I did here](https://github.com/flybywiresim/a32nx/pull/2262/files).

**Comment by saintwolf** [comment]
Added spaces, and updated some other text as per the FMS reference:

image
image

**Comment by beheh** [comment]
Two more things: From my last reference, the selected items are yellow while in a tempy flight plan (when INSERT/DELETE is shown) and the header could be updated to say `ARRIVAL {small}TO{end} {green}EDDF{end}`

**Comment by saintwolf** [comment]
@beheh How's this look?

image
image
image
image
image

**Comment by saintwolf** [comment]
Okay I think I'm done with all of the changes. @beheh @Benjozork Can you take a look at the updated code please?

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AvailableArrivalsPage.js [hunk 2]
```diff
@@ -5,38 +5,68 @@ class CDUAvailableArrivalsPage {
...
+                    selectedTransitionCellColor = mcdu.flightPlanManager.getCurrentFlightPlanIndex() === 1 ? "yellow" : "green";
                 }
             }
             const approaches = airportInfo.approaches;
+            // Sort the approaches in Honeywell's documented order
+            const sortedApproaches = approaches.sort(
+                function (a, b) {
+                    const approachTypeOrder = {"MLS":0, "ILS":1, "GLS":2, "IGS":3, "LOC":4, "BLOC":5, "LDA":6, "SDF": 7, "GPS": 8, "RNAV":9, "VORDME":10, "NDB":11};
+                    return approachTypeOrder[a.name.split(" ")[0]] - approachTypeOrder[b.name.split(" ")[0]];
+                }
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AvailableArrivalsPage.js:48]: Use an arrow function instead.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AvailableArrivalsPage.js [hunk 3]
```diff
@@ -5,38 +5,68 @@ class CDUAvailableArrivalsPage {
...
-                    const approach = approaches[index];
+                    const approach = sortedApproaches[index];
                     if (approach) {
-                        rows[2 * i] = ["{" + Avionics.Utils.formatRunway(approach.name) + "[color]cyan"];
+                        const runways = airportInfo.oneWayRunways;
+                        const approachRunwayName = Avionics.Utils.formatRunway(approach.name.split(" ")[1]);
+                        let runwayLength = 0;
+                        let runwayCourse = 0;
+                        for (let i = 0; i < runways.length; i++) {
+                            const runway = runways[i];
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AvailableArrivalsPage.js:61]: Use a `for ... of` loop instead.

**Comment by BritishAvgeek** [comment]
Category: QA Tester
Name: BritishAvgeek#1436
Date of testing: Current date 14.12.2020
Version of the sim: 1.11.6.0
PR Tested: #2263 
Object of test: Arrival Page Fixes

Testing technique: Full Flight livestreamed here https://youtu.be/tcRMJ7YzaS0    Mumbai to Delhi
Steps you've made to test the Pull Request:
1. Generated flight plan using SimBrief linked to Navigraph and AIRAC 2013
2. Spawned C&D - Input plan into MCDU as per plan
3. During input of departure, it reverted to it's behaviour pre-fix so the Temp FPLAN and yellow for the departure no longer showed
4. During input of Planned Arrival, it was all off - I had to select ILS 27 for Runway 28 - but it then gave me VOR DME 28 instead of ILS - but still input the ILS freq into the RADNAV (see comments below)
5. Flew route and landed, no issues through flight except the wider known sim issues

Flight notes (If required):
Route: VABB/09 ALBA2B ISRIS DCT DOSTO DCT EGUGU DCT GOMTI DCT INTIX DCT NOKOK DCT BUBNU DCT DUKOT DCT ELMON ATRI5B/28

Flew the VOR DME arc manually anyway on arrival. It wasn't incorrect in forcing the VOR DME 28 but bearing in mind the selection was ILS 27 to trigger the button below which read ILS 28 - it should have displayed ILS 28 instead and fed the RNAV arrival for ATRI5B in, instead of the DME ARC. I flew the ARC anyway for practice and to show people watching the stream how to fly it but the input into the MCDU seemed wrong

Go to:
https://www.youtube.com/watch?v=tcRMJ7YzaS0&feature=youtu.be
Relevant Times:
06mins 10 seconds for Departure Input to show the issues now arising from that - gave us just "ERASE" or "INSERT" as options for the SID - not a return to flight plan option 
10mins 40seconds onwards for Arrival Input issues but I selected ILS28 and it gave me VORDME29 - I selected ILS27 and it gave me VORDME28
2:00:00 onwards for the VOR DME Arrival and ILS approach

Issues (If Any): 
Overall Rating: To be improved and needs further testing
Conclusions of the testing: All mentioned above - unsure as to whether the VORDME read properly on the arrival page but the input was for an ILS approach (which is fed into the MCDU RADNAV page anyway) not a VOR DME which I believe should have been input as the navigraph screenshot attached shows:
image


I ended up flying the DME ARC anyway as shown below as this is also what it was forcing (orange line):
image

**Comment by beheh** [comment]
> 06mins 10 seconds for Departure Input to show the issues now arising from that - gave us just "ERASE" or "INSERT" as options for the SID - not a return to flight plan option

That behavior sounds correct - there should only be ERASE and INSERT, and you can click FPLN if you want to see the flight plan. Pressing INSERT should insert and take you to the flight plan immediately.

image

**Comment by BritishAvgeek** [comment]
What about the rest?
That wasn't as much an issue but an observation of it working differently but the actual arrival runway selection was seemed all wrong with the wrong runways and approaches being chosen from input etc.

**Comment by beheh** [comment]
> Also should the departure selection and pressing Insert not keep it as a temp flight plan as it did before then returning to the FLT PLN page displaying all as yellow/temp? This one committed the SID in green and I had to go in and trigger it into temp Flt PLN by continuing input of the route.

I thought it behaved exactly like this before, but in any case, no: "Insert" immediately overwrites the current, active flight plan. If you want to see a preview you need to explicitly press the FPLN button.

image

**Comment by BritishAvgeek** [comment]
Cheers for that, good to know... Still the arrivals page issues need looking into, as they appeared quite wrong for my destination airport today

**Comment by SierraBravo1** [comment]
The mismatch with the approach selection on the Arrival page is also present for an arrival to EDDN.
The approach selection is shifted by one item.

If I select -> I get:

ILS 10  -> VORDME 10
ILS 28Z -> ILS 10
LOC 28Y -> ILS 28
RNAV 10 -> LOC 28
VORDME 10 -> RNAV 10

**Comment by saintwolf** [comment]
Thanks for the testing folks, I know why this is happening now. Basically I am reordering the array of approaches, however the index of that approach is required to select it afterwards. Does anyone know if this index is stored within the approach object at all? E.g. with something like getApproachIndex()?

**Comment by SierraBravo1** [comment]
hanks, saintwolf.
I found another issue with the arrival page. It would be great if you could have a look at it.

Set EDDS as destination.
Select approach ILS 07
Select STAR DENE1E
Select No VIA

DENE1E arrival is added, but the approach via BADSO is inserted automatically, which cannot be removed.
Both, STAR DENE1E and the approach via BADSO are active in the flightplan.
It looks like it is not possible to select NO VIAS. If done, the BADSO approach is included in addition.

EDDS_DENE1E

**Comment by saintwolf** [comment]
@SierraBravo1 There is no special handling in the menu for the No VIAs option (it's built in), it is likely something to do with Asobo's flight plan manager, which I am unable to fix in this code unfortunately.

**Comment by saintwolf** [comment]
@SierraBravo1 I think this is due to the function call setApproachTransitionIndex(-1) which is not being handled correctly, however it's not something I can fix in the scope of the arrival page code unfortunately. This would need a separate issue/PR to investigate (if we can even edit the flight plan manager code without affecting other aircraft).

**Comment by Awemeter** [comment]
Category: QA Tester
Name: Awemeter#7919
Date of testing: 12/17/2020
Version of the sim: 1.11.7.0
PR Tested: #2263 
Object of test: Arrival pages. 

Testing technique: Full flight on a route that I know has double missing stars.
Steps you've made to test the Pull Request: 
1. C&D Full flight Kord to Kdca. 
2. Programmed flight plan and did not see double stars. I also saw the enhancements on the arrival page.
3. Landed at DCA


Flight notes (If required): 

Issues (If Any): N/A
Overall Rating: Good
Conclusions of the testing: A very great pr that allows me to fly into airports with a messed up arrival page. Thanks for your hard work and keep it up.

**Comment by Benjozork** [comment]
> @SierraBravo1 I think this is due to the function call setApproachTransitionIndex(-1) which is not being handled correctly, however it's not something I can fix in the scope of the arrival page code unfortunately. This would need a separate issue/PR to investigate (if we can even edit the flight plan manager code without affecting other aircraft).

We can, just gotta copy it over. But that's already being worked on by @wpine215. There's a proposal in flybywiresim/rfcs

**Comment by SierraBravo1** [comment]
> 
> 
> Thanks for the testing folks, I know why this is happening now. Basically I am reordering the array of approaches, however the index of that approach is required to select it afterwards. Does anyone know if this index is stored within the approach object at all? E.g. with something like getApproachIndex()?

Fixed for me
Tested approach selection of EDDN and LFMN successfully.
Thanks

**Comment by tjburni** [comment]
Category: QA Tester
Name: [Z+1]tjburni#0409
Date of testing: Current date 19.12.2020
Version of the sim: 1.11.7
PR Tested: #2263
Object of test:Arrival Pages

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Full IFR Flight from C&D 
2. Programmed flight plan and checked for the fixed issues descrobed in the pr
3. Landed
4.
5.

Flight notes (If required):

Issues (If Any): none
Overall Rating: Good
Conclusions of the testing: Looks good for me all issues are gone. Thanks

**Pull request merged**